### ISSUE TITLE
DynamicMapLayer crossing dateline

### ISSUE THREAD

**Issue by rowanwins** [Esri/esri-leaflet#672]
Gday guys,

Im currently doing up a global map and am bringing in a DynamicMapLayer however Im running into the dateline issue where when zoomed out the layer suddenly stops.

L.tilelayer contains an option called [continousWorld](http://leafletjs.com/reference.html#tilelayer-continuousworld) which I think would address my issue however it looks like DynamicMapLayer only inherits the options from L.ImageOverlay .

Do you have any suggestions as to how to best tackle this? I'm happy to take a look at the tilelayer code and see if there is anything usable in there if you think that's as good a starting point as any...

Cheers
Rowan

![Issue](https://www.dropbox.com/s/48m7rw3uft0zmva/dynamicLayer.jpg?dl=1)

PS I wont ask you fix the [vector geometry dateline problem](https://github.com/Leaflet/Leaflet/pull/1293), that looks really complex!

**Comment by rowanwins** [comment]
actually on second thoughts this might be being managed by the noWrap option on tilelayer
http://leafletjs.com/reference.html#tilelayer-nowrap

**Comment by patrickarlt** [comment]
In a perfect world you would probably want this right?

![fullscreen_10_29_15__7_04_am](https://cloud.githubusercontent.com/assets/378557/10820676/b8181aa4-7e0b-11e5-9c47-f2e0b31208ae.png)

I'm not entirely sure a solution to this is possible in Leaflet given all the vector wrapping problems. If there was a good solution to this I would still have to manage multiple images for 1 bounding box which is a HUGE complexity increse. You might consider just setting `noWrap` to true on your tile layer and `maxBounds` on your map http://leafletjs.com/reference.html#map-maxbounds.

**Comment by rowanwins** [comment]
That's a shame @patrickarlt , I'd hoped that the dynamic layer acted more like a tile layer and so the wrapping would be easier but it doesn't sound like that's the case.

I'm currently working on a map with global data (shipping) hence the global interest.

Interestingly it looks like Open Layers is also struggling with this a bit. The [ol.source.TileArcGISRest](http://openlayers.org/en/v3.10.1/apidoc/ol.source.TileArcGISRest.html) has an option for wrapX however I had a quick play and couldn't get that to work either :(

Anyway I'll close the issue but I'll live in hope that maybe one day someone will have a brainwave about how to tackle it :)

**Comment by rowanwins** [comment]
Just as a quick followup for others who are interested. 

Basically a dynamicMapLayer request is only returning a single image from ArcGIS Server. It seems like a limitation of  ArcGIS Server that it doesn't want to cross the dateline boundary, for example
http://sampleserver5.arcgisonline.com/arcgis/rest/services/SampleWorldCities/MapServer/export?bbox=9725235.982779548%2C3757032.8142729844%2C25887944.23584978%2C9666532.34505653&bboxSR=3857&layers=&layerDefs=&size=1652%2C604&imageSR=3857&format=png&transparent=false&dpi=96&time=&layerTimeOptions=&dynamicLayers=&gdbVersion=&mapScale=&f=html

The full image size is returned by you just get a big transparent area on the right of the image. So it's actually nothing to do with leaflet or esri-leaflet. 

So basically you'd need to pass a 2nd request to ArcGIS Server to get the remaining half of the picture.

I kinda feel like its not that outside the realm of possibility @patrickarlt , the main thing Im not sure about is how you'd build a loop into requesting multiple images...

**Comment by jgravois** [comment]
> it looks like Open Layers is also struggling with this a bit

works just fine in the [jsapi](it looks like Open Layers is also struggling with this a bit) :smile:.

**Comment by rowanwins** [comment]
Perhaps that's why the jsapi it takes 50MB to download ;)

**Comment by patrickarlt** [comment]
> Basically a dynamicMapLayer request is only returning a single image from ArcGIS Server. It seems like a limitation of ArcGIS Server that it doesn't want to cross the dateline boundary, for example

I don't think it is this. The dynamic map layer sample in the JS API runs just fine across the date line with a single image http://developersdevext.arcgis.com/javascript/sandbox/sandbox.html?sample=map_dynamic

I'm going to reopen this for investigation.

**Comment by rowanwins** [comment]
Gday @patrickarlt , unfortunantly I can't see the example link you sent (it looks like its locked so that its only available to esri employees). Thanks for taking another look into it, fingers crossed :)

**Comment by patrickarlt** [comment]
@rowanwins whoops http://developers.arcgis.com/javascript/sandbox/sandbox.html?sample=map_dynamic is the correct link

**Comment by rowanwins** [comment]
ah yes, I had looked at that yesterday but im not sure it is achieving what we'd expect either?

![Issue](https://www.dropbox.com/s/qwme79e7peor750/jsapi.jpg?dl=1)

I also came across these examples which I thought might be relevant but they dont seem to tackle the issue either.
https://developers.arcgis.com/javascript/jssamples/ed_wraparound.html
https://developers.arcgis.com/javascript/jssamples/map_wraparound.html

**Comment by jgravois** [comment]
^^ i fixed the link in my previous cheeky comment

**Comment by rowanwins** [comment]
So looking at @jgravois sample above the image requst being sent is 
````
http://sampleserver5.arcgisonline.com/arcgis/rest/services/SampleWorldCities/MapServer/export?
dpi=96
&transparent=true
&format=jpeg
&bbox=-3377905.15397761,6418988.26507475,3377905.15397761,9041084.08336874
&bboxSR={"wkt":"PROJCS[\"WGS_1984_Web_Mercator_Auxiliary_Sphere\",GEOGCS[\"GCS_WGS_1984\",DATUM[\"D_WGS_1984\",SPHEROID[\"WGS_1984\",6378137.0,298.257223563]],PRIMEM[\"Greenwich\",0.0],UNIT[\"Degree\",0.0174532925199433]],PROJECTION[\"Mercator_Auxiliary_Sphere\"],PARAMETER[\"False_Easting\",0.0],PARAMETER[\"False_Northing\",0.0],PARAMETER[\"Central_Meridian\",167.76988636017128],PARAMETER[\"Standard_Parallel_1\",0.0],PARAMETER[\"Auxiliary_Sphere_Type\",0.0],UNIT[\"Meter\",1.0]]"}
&imageSR={"wkt":"PROJCS[\"WGS_1984_Web_Mercator_Auxiliary_Sphere\",GEOGCS[\"GCS_WGS_1984\",DATUM[\"D_WGS_1984\",SPHEROID[\"WGS_1984\",6378137.0,298.257223563]],PRIMEM[\"Greenwich\",0.0],UNIT[\"Degree\",0.0174532925199433]],PROJECTION[\"Mercator_Auxiliary_Sphere\"],PARAMETER[\"False_Easting\",0.0],PARAMETER[\"False_Northing\",0.0],PARAMETER[\"Central_Meridian\",167.76988636017128],PARAMETER[\"Standard_Parallel_1\",0.0],PARAMETER[\"Auxiliary_Sphere_Type\",0.0],UNIT[\"Meter\",1.0]]"}&size=1381,536&f=image
````
It seems like the main critical thing is the custom wkt passed to the imageSR and bboxSR params... Unfortunately I dont know too much about how those wkt strings are structured. Also given that the user can set these as params I'm not how this could be tackled in an easy way.

Anyway thanks for getting us this far @patrickarlt & @jgravois 
Cheers
Rowan

**Comment by patrickarlt** [comment]
@jgravois @rowanwins If you add `basemap: 'topo'` to the JS API example i previous posted it works wrapping around the dateline. I think it has to do with which spatial reference you use.

I'll keep investigating this when I have some time later.

**Comment by molnar** [comment]
@rowanwins,  here is a working example I started working on last week.  I ran into the same issue you are having with data extending passing the date line.

https://jsfiddle.net/u9raxmcx/
Pan to the west.

Modifying EsriLeaflet.Layers.DynamicMapLayer _buildExportParams function should allow custom imageSR and bboxSR such as the jsapi does.

Before the the params are returned, this was added:

     var groundResolution = {      
      "1":78271.5170,
      "2":39135.7585,
      "3":19567.8792,
      "4":9783.9396,
      "5":4891.9698,
      "6":2445.9849,
      "7":1222.9925,
      "8":611.4962,
      "9":305.7481,
      "10":152.8741,
      "11":76.4370,
      "12":38.2185,
      "13":19.1093,
      "14":9.5546,
      "15":4.7773,
      "16":2.3887,
      "17":1.1943,
      "18":0.5972,
      "19":0.2986,
      "20":0.1493,
      "21":0.0746,
      "22":0.0373,
      "23":0.0187
    } //https://msdn.microsoft.com/en-us/library/bb259689.aspx     

    var centralMeridian =  null; 
    var tempbbox = params.bbox.split(','); 
    var goingEast = false;   
    
    if(Math.abs(tempbbox[0]) > 20037507.0671618 || Math.abs(tempbbox[2]) > 20037507.0671618){      
      centralMeridian =  this._map.getCenter().lng;
      if(centralMeridian >= 180) goingEast = true;    
      if(Math.abs(centralMeridian) > 180){       
        centralMeridian = 180 - ((Math.abs(centralMeridian) - 180 ));
      }
      if(goingEast)centralMeridian = -centralMeridian;
      var zoom = this._map.getZoom()+1;
      var screenwidth = this._map.getSize().x;
      tempbbox[0] = -(groundResolution[zoom]*screenwidth);
      tempbbox[2] = (groundResolution[zoom]*screenwidth);
      params.bbox = tempbbox.join(",");
    }
    params.imageSR = {"wkt":"PROJCS[\"WGS_1984_Web_Mercator_Auxiliary_Sphere\",GEOGCS[\"GCS_WGS_1984\",DATUM[\"D_WGS_1984\",SPHEROID[\"WGS_1984\",6378137.0,298.257223563]],PRIMEM[\"Greenwich\",0.0],UNIT[\"Degree\",0.0174532925199433]],PROJECTION[\"Mercator_Auxiliary_Sphere\"],PARAMETER[\"False_Easting\",0.0],PARAMETER[\"False_Northing\",0.0],PARAMETER[\"Central_Meridian\","+centralMeridian+"],PARAMETER[\"Standard_Parallel_1\",0.0],PARAMETER[\"Auxiliary_Sphere_Type\",0.0],UNIT[\"Meter\",1.0]]"};
    params.bboxSR = {"wkt":"PROJCS[\"WGS_1984_Web_Mercator_Auxiliary_Sphere\",GEOGCS[\"GCS_WGS_1984\",DATUM[\"D_WGS_1984\",SPHEROID[\"WGS_1984\",6378137.0,298.257223563]],PRIMEM[\"Greenwich\",0.0],UNIT[\"Degree\",0.0174532925199433]],PROJECTION[\"Mercator_Auxiliary_Sphere\"],PARAMETER[\"False_Easting\",0.0],PARAMETER[\"False_Northing\",0.0],PARAMETER[\"Central_Meridian\","+centralMeridian+"],PARAMETER[\"Standard_Parallel_1\",0.0],PARAMETER[\"Auxiliary_Sphere_Type\",0.0],UNIT[\"Meter\",1.0]]"};

**Comment by patrickarlt** [comment]
@molnar I've been neck deep in other projects for awhile. Could you update your fiddle to work again?

**Comment by molnar** [comment]
@patrickarlt, sorry about that, I will get it up and running this evening

**Comment by molnar** [comment]
Here is an updated fiddle
https://jsfiddle.net/u9raxmcx/1/

**Comment by patrickarlt** [comment]
Thanks for making that demo work again @molnar. I'm going to look into this some more and see if I can work this into the core library.

**Comment by SVoyt** [comment]
Check my solution https://github.com/SVoyt/esri-leaflet/tree/feature-advanced-dynamic.
I reworked DynamicMapLayer ( `L.esri.dynamicMapLayerAdvanced` ),
This layer's class is independent from DynamicMapLayer and RasterLayer and doesn't use recreating ImageOverlay on each update (so order of layers is permanent). It works properly with dateline, world repeating and zIndex.
On this week i will finish it ( need to fix some bugs and write some tests).

**Comment by jgravois** [comment]
> It works properly with dateline, world repeating and zIndex.

@SVoyt fantastic!  i'm really looking forward to reviewing your PR.

**Comment by patrickarlt** [comment]
@SVoyt I've got to echo @jgravois here, that sounds really fantastic. Can't wait to review it.

**Comment by SVoyt** [comment]
You can try it [here](http://svoyt.github.io/labs/esri-leaflet/dynamic-advanced/index.html).
It works fine when `f:'image'`, but when `f:'json'` we have bugs. I trying to fix it now and i trying to write right tests.

**Comment by rowanwins** [comment]
Great work @SVoyt , this has come at a very useful time for me!

I did notice there seems to be some empty pixels on the dateline, is this what everyone else see's as well or is this a browser bug? Or is it expected service from the service that you're using?

Looks great otherwise!
Cheers
Rowan

![empty pixels](https://www.dropbox.com/s/audk94r4pthra5k/screen.jpg?dl=1)

**Comment by SVoyt** [comment]
@rowanwins, this grey line is from service, if you'll switch to ocean layer, border will be more smooth. But sometimes there is round-off error, when two images has space between ( ± 1px ), now i'm working on it.

p.s. I fixed bug with `f:json`

**Comment by SVoyt** [comment]
So, I formed it as a plugin for esri-leaflet - [esri-leaflet-dynamic-advanced](https://github.com/SVoyt/esri-leaflet-dynamic-advanced)

**Comment by rowanwins** [comment]
Nice work @SVoyt , I'll give it a go over the next few days as I'm currently working with some global marine data.

From my quick look at the code it doesnt look like you're extending the dynamic layer from the core esri-leaflet library and I wonder if this is a problem? Perhaps you can somehow extend it rather than recreating it... @jgravois might have some advice about how to best structure this as a plugin, it may well be that you're perfectly on track and nothing needs to change!

Cheers

**Comment by jgravois** [comment]
@SVoyt's plugin looks great to me so i added it to our [plugin page](http://esri.github.io/esri-leaflet/plugins/).

i'm going to go ahead and close this issue, but if anyone has more questions, i'd be happy to discuss them.

**Comment by scaddenp** [comment]
This is old but I am still running foul of this on maps around SW pacific that cross dateline. Did the solution ever get wrapped into main code or should I be looking at the dateline (and I am using f:json which is easily enough changed)

**Comment by jgravois** [comment]
> Did the solution ever get wrapped into main code

it didn't.

**Comment by scaddenp** [comment]
Pity. I have got the plugin and it is working fine (phew). Back to the problem of spatial querying with a selection polygon that cross the antemeridian (or include a pole).
